TypedSyntax/src/node.jl

```diff
@@ -0,0 +1,339 @@
+#     clidx = searchsortedfirst(src.linetable, lineno; lt=(linenode, line) -> linenode.line < line)
+#     clidx = min(clidx, length(src.linetable))
+#     if src.linetable[clidx].line > lineno
+#         clidx -= 1    # handle multiline statements
+#     end
+#     return clidx
+# end
+# function find_coderange(src, lineno)
+#     clidx = find_codeloc(src, lineno)
+#     ibegin = searchsortedfirst(src.codelocs, clidx)
+#     ibegin += src.codelocs[ibegin] > lineno
+#     iend = searchsortedlast(src.codelocs, clidx)
+#     return ibegin:iend
+# end
+
+function sparam_name(mi::MethodInstance, i::Int)
+    sig = (mi.def::Method).sig::UnionAll
+    while true
+        i == 1 && break
+        sig = sig.body::UnionAll
+        i -= 1
+    end
+    return sig.var.name
+end
+
+# # Get the name of the function being called
+# function extract_call_name(stmt, mi)::Union{Symbol,Nothing}
+#     stmt.head ∈ (:call, :invoke) || return nothing
+#     f = stmt.args[1]
+#     if isa(f, GlobalRef) && f.mod === Core && f.name == :_apply_iterate   # handle vararg calls
+#         # Sanity check
+#         fiter = stmt.args[2]
+#         @assert isa(fiter, GlobalRef) && fiter.name == :iterate
+#         f = stmt.args[3]   # get the actual call
+#     end
+#     if isa(f, SlotNumber)
+#         return src.slotnames[f.id]
+#     end
+#     if isa(f, SSAValue)
+#         f = src.ssavaluetypes[f.id]
+#         if isa(fname, Core.Const)
+#             f = fname.val
+#         end
+#     end
+#     if isexpr(f, :static_parameter)
+#         return sparam_name(mi, f.args[1]::Int)
+#     end
+#     if isa(f, QuoteNode)
+#         f = f.val
+#     end
+#     isa(f, GlobalRef) && return f.name
+#     return nothing
+# end
+
+function getsrc(@nospecialize(f), @nospecialize(t))
+    srcrts = code_typed(f, t; debuginfo=:source, optimize=false)
+    return only(srcrts)
+end
+
+function is_function_def(node)  # this is not `Base.is_function_def`
+    kind(node) == K"function" && return true
+    kind(node) == K"=" && kind(child(node, 1)) == K"call" && return true
+    return false
+end
+
+function collect_symbol_nodes!(symlocs::AbstractDict{Symbol}, node)
+    kind(node) == K"->" && return symlocs     # skip inner functions (including `do` blocks below)
+    is_function_def(node) && return symlocs
+    if kind(node) == K"Identifier"
+        name = node.val
+        if isa(name, Symbol)
+            locs = get!(Vector{typeof(node)}, symlocs, name)
+            push!(locs, node)
+        end
+    elseif haschildren(node)
+        for c in (kind(node) == K"do" ? (child(node, 1),) : children(node))  # process only `g(args...)` in `g(args...) do ... end`
+            collect_symbol_nodes!(symlocs, c)
+        end
+    end
+    return symlocs
+end
+
+# Find all places in the source code where a symbol is used
+function collect_symbol_nodes(rootnode)
+    kind(rootnode) ∈ KSet"function =" || error("expected function definition, got ", sourcetext(kind(rootnode)))
```

Comment:
I think we need to register new version of JuliaSyntax so that TypedSyntax can use it.
After `dev`ing JuliaSyntax to the latest master, I tried `TypedSyntaxNode` for my test case `TypedSyntaxNode(sin, (Int,))` and I got error here. It looks like `rootnode` can be `@eval`ed function definition and so we want to handle it here somehow?

Reply:
A new release would be good. I also forgot I had a couple more local modifications, you may need https://github.com/JuliaLang/JuliaSyntax.jl/pull/208. But your example was useful, fixed in the latest commit.

Reply:
BTW, you've probably noticed this already, but I deliberately placed a bunch of `@assert`s in lots of places. The new design does involve understanding the code at a deeper level, and the `@assert`s are intended to make it fail as soon as something unexpected arises. It's going to be very fragile in these early days, but deliberately so. Feel free to just list test cases that fail and I will address them.